### PR TITLE
Add pytorch-notebook image variants with cuda 11 and 12 (x86_64 versions only)

### PR DESCRIPTION
**Pull request by johanna-reiml-hpi**
## Describe your changes

This reworks the build pipeline to allow building CUDA variants for the `pytorch-notebook` image. Examples images: https://quay.io/repository/ai-hpi/pytorch-notebook

- **Commit** "feat: build cuda variants of pytorch" (43f2d320): minimal example for how to add cuda variants for the pytorch-notebook image.
- **Commit** "feat: build with variant tag" (6198575d): rework the pipeline to build these images under the tag of the `pytorch-notebook` instead (adds a "cuda-" and "cuda11-" prefix). This logic is implemented by adding a new "variant" argument for the GitHub Actions and the tagging scripts. I also refactored `docker.yaml` to get rid of the redundancy for the aarch64 vs x86_64 build.

In the [issue](https://github.com/jupyter/docker-stacks/issues/1557) there were discussions about using the [official CUDA images](https://catalog.ngc.nvidia.com/orgs/nvidia/containers/cuda/tags) instead of `ubuntu:22.04` as the root container. This is not desirable in my opinion, as these images are not updated frequently and are based on older `ubuntu:22.04` images. Instead a simpler approach works: packages like pytorch and tensorflow bundle their own CUDA binaries via pip these days.

I looked into CUDA builds for TensorFlow and PyTorch (I am not sure if CUDA support is really a priority for other notebooks). In order to [support NVIDIA driver versions](https://docs.nvidia.com/deploy/cuda-compatibility/) older than 525, builds for CUDA 11.8 should also be included.

- For PyTorch (https://pytorch.org/get-started/locally/) everything works as expected. I also added NVIDIA's pip index for their most recent packages as an install argument `--extra-index-url=https://pypi.nvidia.com`, but it's likely not needed.
- For TensorFlow (https://www.tensorflow.org/install/source?hl=en#gpu) there were multiple issues:
    - CUDA 11: The most recent version (2.15.0) updated the CUDA dependencies from 11.8 to 12.2. So the last version which supports CUDA 11.8 is 2.14.0. However, that version is not functional under python 3.11, because a dependency (tensorrt==8.5.3.1) is not available for python 3.11: https://github.com/tensorflow/tensorflow/issues/61986
    - CUDA 12 (latest): Using `pip install --no-cache-dir --extra-index-url=https://pypi.nvidia.com tensorrt tensorflow[and-cuda]` didn't seem to work for me and tensorflow was [unable to find tensorrt](https://github.com/tensorflow/tensorflow/issues/61468). I assume the workaround in the linked issue would work, but it seems better to either manually build the package, try nightly builds, or wait for 2.16.0.

## Issue ticket if applicable

Fix (for pytorch): https://github.com/jupyter/docker-stacks/issues/1557

## Checklist (especially for first-time contributors)

- [x] I have performed a self-review of my code
- [x] If it is a core feature, I have added thorough tests
    - [x] Example build: https://github.com/johanna-reiml-hpi/docker-stacks/actions/runs/7788922093
    - [x] Example images: https://quay.io/repository/ai-hpi/pytorch-notebook?tab=tags
    - [x] No aarch64 tests yet (I assume you want to use your self-hosted runners)
- [x] I will try not to use force-push to make the review process easier for reviewers
- [x] I have updated the documentation for significant changes
    - [x] Not sure what parts need to be updated. I mentioned the existence of optional CUDA tags under the pytorch-notebook section.

### CODE REVIEW

**Comment by johanna-reiml-hpi** [comment]
From what I can see the build for aarch64 seems to fail due to running out of disk space. Is there an issue with also enabling this part for the self-hosted aarch64 runners? Right now it only runs for x86_64.

```yaml
      # Image with CUDA needs extra disk space
      - name: Free disk space 🧹
        if: contains(inputs.variant, 'cuda') && inputs.platform == 'x86_64'
        uses: jlumbroso/free-disk-space@54081f138730dfa15788a46383842cd2f914a1be
        with:
          tool-cache: false
          android: true
          dotnet: true
          haskell: true
          large-packages: false
          docker-images: false
          swap-storage: false
```

File: tagging/get_prefix.py
Comment (mathbunnyru) [inline tagging/get_prefix.py:15]:
```suggestion
def get_file_prefix_for_platform(platform: str, variant: str) -> str:
    return f"{platform}-{variant}"


def get_tag_prefix_for_platform(platform: str, variant: str) -> str:
    if variant == DEFAULT_VARIANT:
        return platform
    return f"{platform}-{variant}"
```

File: docs/using/selecting.md
Comment (mathbunnyru) [inline docs/using/selecting.md:198]: I think we should add a note section, which says that you should add a `cuda-` or `cuda11-` prefix to the tag to get GPU-enabled versions and explain the difference.

File: Makefile
Comment (mathbunnyru) [reply Makefile:70]: There are always 3 empty lines between sections, let's keep it this way for consistency.

File: .github/workflows/docker-build-test-upload-all.yml
Comment (mathbunnyru) [inline .github/workflows/docker-build-test-upload-all.yml:1]:
```suggestion
name: Build, test, and push Docker Images for one platform
```

File: .github/workflows/docker-build-test-upload-all.yml
Comment (mathbunnyru) [inline .github/workflows/docker-build-test-upload-all.yml:14]:
```suggestion
      fast-runner-tag:
```

File: .github/workflows/docker-build-test-upload-all.yml
Comment (mathbunnyru) [inline .github/workflows/docker-build-test-upload-all.yml:18]:
```suggestion
      slow-runner-tag:
```

File: .github/workflows/docker-build-test-upload-all.yml
Comment (mathbunnyru) [inline .github/workflows/docker-build-test-upload-all.yml:1]: I would have preferred having this refactoring in a separate PR first, but I'm ok keeping it here.
I have another concern though  - can we make left side of the run better?
With this PR, I need to click all build to find `datascience` build, for example.

Screenshot 2024-02-08 at 13 25 08

File: .github/workflows/docker.yml
Comment (mathbunnyru) [inline .github/workflows/docker.yml:3]: We need to move comment about `[FAST_BUILD]` to a new file.

File: .github/workflows/docker.yml
Comment (mathbunnyru) [inline .github/workflows/docker.yml:117]: Could you please explain why this need to this complicated?

**Review by mathbunnyru — CHANGES_REQUESTED** [review]

**Comment by mathbunnyru** [comment]
> there an issue with also enabling this part for the self-hosted aarch64 runners?

Self-hosted runners are not usually bloated with lots of software. I created our aarch64 runners using ubuntu provided by Google Cloud and running this on top: https://github.com/jupyter/docker-stacks/blob/main/aarch64-runner/setup.sh

Could you please tell me how much free disk is needed to build one GPU image?
I am not sure people need aarch64 GPU images, so let's not build them in this PR, though.
Also, this will increase the build time (we have limited amount of self-hosted runners)

File: .github/workflows/docker-build-test-upload-all.yml
Comment (mathbunnyru) [reply .github/workflows/docker-build-test-upload-all.yml:1]: I also think you can still create a separate PR for this change (having `docker-build-test-upload-all.yml` file), then this PR will look much more readable.

File: Makefile
Comment (mathbunnyru) [inline Makefile:39]: Slightly unrelated I suppose (I would be happy to merge this in a separate PR).

File: images/pytorch-notebook/Dockerfile
Comment (mathbunnyru) [inline images/pytorch-notebook/Dockerfile:14]: Thank you for adding a link

File: images/pytorch-notebook/cuda/Dockerfile
Comment (mathbunnyru) [inline images/pytorch-notebook/cuda/Dockerfile:16]: One more thought - let's call this image cuda12 and I think we can add something like `--index-url 'https://download.pytorch.org/whl/cu121'` to the installation command.

When cuda13 will be released in the future, we won't break anyone by silently updating their image from cuda12 to cuda13. We will just add a new prefix tag `cuda13` and everyone who can and wants to upgrade will be able to do that with a small change.

File: .github/workflows/docker-build-test-upload.yml
Comment (mathbunnyru) [inline .github/workflows/docker-build-test-upload.yml:17]: I suggest making `default` the default value here 😃.
Currently, parent image is always default, so we will add the option if someone needs it in the future, but `docker-build-test-upload-all.yml` will be slightly more simple.

**Comment by mathbunnyru** [comment]
@johanna-reiml-hpi I think the quality of this PR is awesome 👍 
Great job.
I left a few comments of some things that may be improved, please, take a look.

We now have a [policy](https://jupyter-docker-stacks.readthedocs.io/en/latest/maintaining/new-images-and-packages-policy.html) for merging new images and adding new packages 🎉

@mathbunnyru, @consideRatio, @yuvipanda, and @manics, please vote 👍 to accept this change and 👎 not to accept it (use a reaction to this message)
The voting deadline is the 8th of March (a month since I posted this message).
The change is accepted, if there are at least 2 positive votes.

We can have a discussion until the deadline, so please express your opinions.

File: .github/workflows/docker.yml
Comment (johanna-reiml-hpi) [reply .github/workflows/docker.yml:117]: This is technically not needed for this repo. Should I remove it?

I added this for my own image builds, because skipped aarch64 builds do not count for `needs: [build-aarch64, build-x86_64]`. If this is skipped it will only push tags with the `x86_64_` prefix. So it will not have tags like `:hub-4.0.2` and only `x86_64-hub-4.0.2`.

Comment (mathbunnyru) [reply .github/workflows/docker.yml:117]: I see. No, I would like my fork to conitnue to work properly (you can see that sometimes I run some things only for `jupyter` GitHub user).
And you might have noticed that your fork works as it should too 🙂

File: .github/workflows/docker-build-test-upload.yml
Comment (johanna-reiml-hpi) [reply .github/workflows/docker-build-test-upload.yml]: Added `default` as the default value.

File: images/pytorch-notebook/cuda/Dockerfile
Comment (johanna-reiml-hpi) [reply images/pytorch-notebook/cuda/Dockerfile]: I added `--index-url 'https://download.pytorch.org/whl/cu121'` and renamed the `cuda` variant to `cuda12`.

File: Makefile
Comment (johanna-reiml-hpi) [reply Makefile]: Reverted.

File: .github/workflows/docker.yml
Comment (johanna-reiml-hpi) [reply .github/workflows/docker.yml]: Added this adjusted check to both the `merge-tags-fast` and `merge-tags` (without the refactored ´-all.yaml`).

File: .github/workflows/docker.yml
Comment (johanna-reiml-hpi) [reply .github/workflows/docker.yml]: Reverted to the original workflow, so let's do this in another PR.

File: .github/workflows/docker-build-test-upload.yml
Comment (johanna-reiml-hpi) [reply .github/workflows/docker-build-test-upload.yml]: Added the ternary operator.

File: .github/workflows/docker-build-test-upload-all.yml
Comment (johanna-reiml-hpi) [reply .github/workflows/docker-build-test-upload-all.yml]: I reverted the refactor.

File: .github/workflows/docker-build-test-upload-all.yml
Comment (johanna-reiml-hpi) [reply .github/workflows/docker-build-test-upload-all.yml]: Reverted the refactor.

File: .github/workflows/docker-build-test-upload-all.yml
Comment (johanna-reiml-hpi) [reply .github/workflows/docker-build-test-upload-all.yml]: Reverted the refactor.

File: .github/workflows/docker-build-test-upload-all.yml
Comment (johanna-reiml-hpi) [reply .github/workflows/docker-build-test-upload-all.yml]: Reverted the refactor.

File: .github/workflows/docker-build-test-upload-all.yml
Comment (johanna-reiml-hpi) [reply .github/workflows/docker-build-test-upload-all.yml]: Reverted the refactor.

File: Makefile
Comment (johanna-reiml-hpi) [reply Makefile:70]: Fixed.

File: tagging/get_prefix.py
Comment (johanna-reiml-hpi) [reply tagging/get_prefix.py]: Adjusted.

File: docs/using/selecting.md
Comment (johanna-reiml-hpi) [reply docs/using/selecting.md]: Added:

> **GPU Acceleration:** Append a CUDA version prefix (`cuda12-` for CUDA 12 or `cuda11-` for CUDA 11) to the image tag to allow PyTorch operations to use compatible NVIDIA GPUs for accelerated computation.

**Comment by johanna-reiml-hpi** [comment]
> > there an issue with also enabling this part for the self-hosted aarch64 runners?
> 
> Self-hosted runners are not usually bloated with lots of software. I created our aarch64 runners using ubuntu provided by Google Cloud and running this on top: https://github.com/jupyter/docker-stacks/blob/main/aarch64-runner/setup.sh
> 
> Could you please tell me how much free disk is needed to build one GPU image? I am not sure people need aarch64 GPU images, so let's not build them in this PR, though. Also, this will increase the build time (we have limited amount of self-hosted runners)

I am not sure how much space is exactly needed, but the [14 GB](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-private-repositories) provided for `ubuntu-latest` weren't enough and resulted in failed builds. I assume probably like 20 GB in total would be sufficient.

File: Makefile
Comment (mathbunnyru) [reply Makefile]: Added this to main as a separate commit.

File: .github/workflows/docker-build-test-upload.yml
Comment (mathbunnyru) [inline .github/workflows/docker-build-test-upload.yml:26]: I suggest making default the default value here as well.

File: .github/workflows/docker.yml
Comment (mathbunnyru) [inline .github/workflows/docker.yml:68]: If you apply the suggestion above, then you won't need to specify `variant: default` here in most of the places.

File: .github/workflows/docker.yml
Comment (mathbunnyru) [inline .github/workflows/docker.yml:367]: My personal preference is to have these three sorted: `default`, `cuda11`, `cuda12`.
In other places as well.

File: tagging/apply_tags.py
Comment (mathbunnyru) [inline tagging/apply_tags.py:78]:
```suggestion
```

Not needed here, it will be string automatically, matches `--owner`.

**Review by mathbunnyru — COMMENTED** [review]
Only small comments. If you have some time, let's change these small details as well.

**Comment by mathbunnyru** [comment]
> > > there an issue with also enabling this part for the self-hosted aarch64 runners?
> > 
> > 
> > Self-hosted runners are not usually bloated with lots of software. I created our aarch64 runners using ubuntu provided by Google Cloud and running this on top: https://github.com/jupyter/docker-stacks/blob/main/aarch64-runner/setup.sh
> > Could you please tell me how much free disk is needed to build one GPU image? I am not sure people need aarch64 GPU images, so let's not build them in this PR, though. Also, this will increase the build time (we have limited amount of self-hosted runners)
> 
> I am not sure how much space is exactly needed, but the [14 GB](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-private-repositories) provided for `ubuntu-latest` weren't enough and resulted in failed builds. I assume probably like 20 GB in total would be sufficient.

I cleaned up the space on our aarch64 runners.
Unfortunately, Docker has an issue, where cleaning up doesn't remove everything it should, so had to clean up manually.

**Comment by mathbunnyru** [comment]
> > > > there an issue with also enabling this part for the self-hosted aarch64 runners?
> > > 
> > > 
> > > Self-hosted runners are not usually bloated with lots of software. I created our aarch64 runners using ubuntu provided by Google Cloud and running this on top: https://github.com/jupyter/docker-stacks/blob/main/aarch64-runner/setup.sh
> > > Could you please tell me how much free disk is needed to build one GPU image? I am not sure people need aarch64 GPU images, so let's not build them in this PR, though. Also, this will increase the build time (we have limited amount of self-hosted runners)
> > 
> > 
> > I am not sure how much space is exactly needed, but the [14 GB](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-private-repositories) provided for `ubuntu-latest` weren't enough and resulted in failed builds. I assume probably like 20 GB in total would be sufficient.
> 
> I cleaned up the space on our aarch64 runners. Unfortunately, Docker has an issue, where cleaning up doesn't remove everything it should, so had to clean up manually.

Seems to work - the build is green 🎉

**Comment by mathbunnyru** [comment]
One more thought - I'm not sure about coda versioning, but it might be worth it to add more precise version of cuda as a tag prefix.
People might want to (or be forced to) stay on 12.1 even when 12.2 `pytorch-cuda` is released.

So, the prefix tag will be `cuda12-` (nothing to change here), but the final image can look something like `jupyter/pythorch-notebook:cuda12-cuda12.1`, so people will be to more precisely say they cuda version.

I don't consider this to be a blocker for this PR, but it would definitely be nice to add this tag later.

**Comment by johanna-reiml-hpi** [comment]
I had another look at the official [pytorch Dockerfile](https://github.com/pytorch/pytorch/blob/main/Dockerfile), [CUDA Dockerfiles](https://gitlab.com/nvidia/container-images/cuda/-/tree/master/dist/12.1.1/ubuntu2204?ref_type=heads) and [nvidia documentation for some env variables](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/1.14.5/docker-specialized.html#dockerfiles), so here are a few extra points to consider:

- The following ENV should be added to the Dockerfile according to the linked NVIDIA documentation:
```
ENV NVIDIA_VISIBLE_DEVICES all
ENV NVIDIA_DRIVER_CAPABILITIES compute,utility
```
- I don't think the CUDA installation for aarch64 worked actually. The CUDA variant build times are identical to the CPU ones. I checked the [official install guide](https://pytorch.org/get-started/locally/) and when selecting MacOS you get: `# CUDA is not available on MacOS, please use default package`. The official pytorch Dockerfile also only installs the CPU version for aarch64. I will exclude aarch64 from this PR.
- The official pytorch Dockerfile installs the CUDA dependencies via conda. This adds CUDA binaries to conventional paths instead of having CUDA as a pip package (`ENV LD_LIBRARY_PATH /usr/local/nvidia/lib:/usr/local/nvidia/lib64`, `ENV PATH /usr/local/nvidia/bin:/usr/local/cuda/bin:$PATH`). I am not sure which approach is better. The current version might be nicer in terms of simplicity (and follows the official install guide).

**Comment by johanna-reiml-hpi** [comment]
> One more thought - I don't know a lot about cuda versioning, but it might be worth it to add more precise version of cuda as a tag prefix (so, we need to add a new tagger to cuda flavored images). People might want to (or be forced to) stay on 12.1 even when 12.2 `pytorch-cuda` is released.
> 
> So, the prefix tag will stay `cuda12-` (nothing to change here), but the final image can look something like `jupyter/pytorch-notebook:cuda12-cuda12.1`, so people will be to more precisely say they cuda version.
> 
> What do you think? I don't consider this to be a blocker for this PR, but it would definitely be nice to add this tag later.

Seems okay to add. However, it's also possible that pytorch might add a ` --index-url 'https://download.pytorch.org/whl/cu122'` alongside the current ` --index-url 'https://download.pytorch.org/whl/cu121'`. In that case image variants would have to be built for both versions. I think it might be less confusing to do something like this instead:

-  `jupyter/pytorch-notebook:cuda12` (latest, so cuda12.2)
-  `jupyter/pytorch-notebook:cuda12.2`
-  `jupyter/pytorch-notebook:cuda12.1`

File: .github/workflows/docker.yml
Comment (johanna-reiml-hpi) [reply .github/workflows/docker.yml:367]: Reordered it.

File: .github/workflows/docker.yml
Comment (johanna-reiml-hpi) [reply .github/workflows/docker.yml]: Added the default.

File: .github/workflows/docker-build-test-upload.yml
Comment (johanna-reiml-hpi) [reply .github/workflows/docker-build-test-upload.yml]: Added the default.

File: tagging/apply_tags.py
Comment (johanna-reiml-hpi) [reply tagging/apply_tags.py]: Fixed.

**Comment by mathbunnyru** [comment]
> > One more thought - I don't know a lot about cuda versioning, but it might be worth it to add more precise version of cuda as a tag prefix (so, we need to add a new tagger to cuda flavored images). People might want to (or be forced to) stay on 12.1 even when 12.2 `pytorch-cuda` is released.
> > So, the prefix tag will stay `cuda12-` (nothing to change here), but the final image can look something like `jupyter/pytorch-notebook:cuda12-cuda12.1`, so people will be to more precisely say they cuda version.
> > What do you think? I don't consider this to be a blocker for this PR, but it would definitely be nice to add this tag later.
> 
> Seems okay to add. However, it's also possible that pytorch might add a ` --index-url 'https://download.pytorch.org/whl/cu122'` alongside the current ` --index-url 'https://download.pytorch.org/whl/cu121'`. In that case image variants would have to be built for both versions. I think it might be less confusing to do something like this instead:
> 
> * `jupyter/pytorch-notebook:cuda12` (latest, so cuda12.2)
> * `jupyter/pytorch-notebook:cuda12.2`
> * `jupyter/pytorch-notebook:cuda12.1`

I think having a separate image for each `major.minor` version of cuda is too much.
We usually only support latest versions.
In case of cuda it makes sense to support both `cuda11` and `cuda12`, but having to support cuda12.Y is too much.
That's why I prefer a solution with a new tagger and a prefix tag as I mentioned above.

File: images/pytorch-notebook/cuda11/Dockerfile
Comment (mathbunnyru) [inline images/pytorch-notebook/cuda11/Dockerfile:23]:
```suggestion
# https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/docker-specialized.html#dockerfiles
```

File: images/pytorch-notebook/cuda12/Dockerfile
Comment (mathbunnyru) [inline images/pytorch-notebook/cuda12/Dockerfile:23]:
```suggestion
# https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/docker-specialized.html#dockerfiles
```

**Review by mathbunnyru — APPROVED** [review]

**Review by mathbunnyru — APPROVED** [review]

**Comment by mathbunnyru** [comment]
@consideRatio @manics please, vote here: https://github.com/jupyter/docker-stacks/pull/2091#issuecomment-1934654952

I am ready to merge this before the vote deadline if I receive one more positive vote.

**Comment by mathbunnyru** [comment]
> Vote to drop CUDA ambitions for tensorflow-notebook

If someone manages to enable cuda-enabled tensorflow image for our docker stacks in a sane way (installing pip/mamba packages, no manual build), then it means that the ecosystem is mature enough for us to support it.
I won't be against merging such a change.
In any case, after merging this PR, it should be easy to experiment.

I mostly agree with @consideRatio's "Misc PR review" section - @johanna-reiml-hpi please, take a look.

> Supported architectures

👍 

> Variant description

I prefer not to change anything here (I really like all the `cuda11` image tags have a long common prefix like `jupyter/pytorch-notebook:cuda11-`)

> PR Title

👍

> Cuda policy

👍

**Comment by consideRatio** [comment]
Ah, i thought it was a suffix seeing f"{platform}-{variant}" somewhere, and that it should be seen as a suffix being put last there. But I think i misunderstood this and that its really a high prio prefix in the tag, so it makes sense to reference it as a prefix still.

I think the description could be improved on still though as i struggled to piece together things. Now i think something like "Some images has variants, they will be published with the same image name, but get a prefix added to the tag." could have been helpful to read somewhere.

**Comment by mathbunnyru** [comment]
I am merging this one - I will add a policy and some documentation.

**Comment by mathbunnyru** [comment]
Done: https://github.com/jupyter/docker-stacks/pull/2097/files

**Comment by mathbunnyru** [comment]
New images are pushed and ready to use 🎉 
https://quay.io/repository/jupyter/pytorch-notebook?tab=tags

**Comment by mathbunnyru** [comment]
@johanna-reiml-hpi thank you for this PR.
It's nice to have GPU enabled image at least for pytorch and I think `variant` feature might be used in other circumstances.

**Pull request merged**
